basic_autonomy_ros2/src/basic_autonomy.cpp

```diff
@@ -0,0 +1,1113 @@
+                throw std::invalid_argument("Accel limits should be positive");
+            }
+
+            bool optimize = true;
+            std::unordered_set<size_t> visited_idx;
+            visited_idx.reserve(curv_speeds.size());
+
+            std::vector<double> output = curv_speeds;
+
+            while (optimize)
+            {
+                auto min_pair = min_with_exclusions(curv_speeds, visited_idx);
+                int min_idx = std::get<1>(min_pair);
+                if (min_idx == -1)
+                {
+                    break;
+                }
+
+                visited_idx.insert(min_idx); // Mark this point as visited
+
+                double v_i = std::get<0>(min_pair);
+                double x_i = downtracks[min_idx];
+                for (int i = min_idx - 1; i > 0; i--)
+                {   // NOTE: Do not use size_t for i type here as -- with > 0 will result in overflow
+                    //       First point's speed is left unchanged as it is current speed of the vehicle
+                    double v_f = curv_speeds[i];
+                    double dv = v_f - v_i;
+
+                    double x_f = downtracks[i];
+                    double dx = x_f - x_i;
+
+                    if (dv > 0)
+                    {
+                        v_f = std::min(v_f, sqrt(v_i * v_i - 2 * accel_limit * dx)); // inverting accel as we are only visiting deceleration case
+                        visited_idx.insert(i);
+                    }
+                    else if (dv < 0)
+                    {
+                        break;
+                    }
+                    output[i] = v_f;
+                    v_i = v_f;
+                    x_i = x_f;
+                }
+            }
+
+            log::printDoublesPerLineWithPrefix("only_reverse[i]: ", output);
+
+            output = trajectory_utils::apply_accel_limits_by_distance(downtracks, output, accel_limit, accel_limit);
+            log::printDoublesPerLineWithPrefix("after_forward[i]: ", output);
+
+            return output;
+        }
+
+        std::vector<carma_planning_msgs::msg::TrajectoryPlanPoint> trajectory_from_points_times_orientations(
+            const std::vector<lanelet::BasicPoint2d> &points, const std::vector<double> &times, const std::vector<double> &yaws,
+            rclcpp::Time startTime)
+        {
+            if (points.size() != times.size() || points.size() != yaws.size())
+            {
+                throw std::invalid_argument("All input vectors must have the same size");
+            }
+
+            std::vector<carma_planning_msgs::msg::TrajectoryPlanPoint> traj;
+            traj.reserve(points.size());
+
+            for (size_t i = 0; i < points.size(); i++)
+            {
+                carma_planning_msgs::msg::TrajectoryPlanPoint tpp;
+                rclcpp::Duration relative_time(times[i] * 1e9); // Conversion of times[i] from seconds to nanoseconds
+                tpp.target_time = startTime + relative_time;
```

Comment:
Curious about this one, does this compile? Can you directly assign rclcpp::Time to builtin_interfaces/Time